### PR TITLE
Updated letter-shop tutorial

### PR DESCRIPTION
**Pull request by adrianhopebailie**
This is a pretty big re-write of @michielbdejong 's awesome tutorial

It teaches the same things and uses the same code. It changes the approach to be more hands on (user completes the code piece by piece) and adds a lot more commentary and logging.

From here users will be ready to move onto tutorials that teach multi-hop (and peer communication using `sendRequest`) or derived protocols like HTTP-ILP.

### CODE REVIEW

**Comment by michielbdejong** [comment]
When you cd into this repo on this branch and run `npm install; npm test` you see a lot of lint errors, mostly about spacing and indentation. Would be good to fix those, if possible.

File: letter-shop/index.md [hunk 1]
```diff
@@ -0,0 +1,634 @@
...
+The server will perform two functions:
+1. It will host an HTTP service where clients can GET a new random letter with each request.
+1. It will listen for incoming ILP payments related to pending letter purchases, fulfill these to accept the payment, and provide the payer with a token they can use to claim their letter.
+
+Once we have completed the tutorial the server will perform the following steps:
+
+1. Connect to an account and monitor it for incoming payments
+1. Start a web server to accept letter requests
+1. Process a request, store the result, and provide the customer with details of how to pay for the request and get the result
+1. Process an incoming payment related to a previous request, and provide payer with details of how to perform request again using proof-of-payment token
```
Comment (michielbdejong) [inline letter-shop/index.md]: the server doesn't provide the payer with details of how to perform [the] request again, it just provides the proof-of-payment token (i.e. the fulfillment). The client then composes the instructions of how to get the letter using that fulfillment.

File: letter-shop/index.md [hunk 2]
```diff
@@ -0,0 +1,634 @@
...
+1. Process a request with an attached proof-of-payment token, and return the originally generated result
```
Comment (michielbdejong) [inline letter-shop/index.md]: nit: the token is not really 'attached', it's 'in the URL'

File: letter-shop/index.md [hunk 3]
```diff
@@ -0,0 +1,634 @@
...
+
+### Client
+
+The client will perform an ILP payment and output the returned fulfillment. This can be redeemed to get the letter that was requested in the original request.
+
+Once we have completed the tutorial the client will perform the following steps:
+
+1. Connect to an account from which to send payments
+1. Initiate a payment using the address, amount, and condition provided in the response to the original letter request
+1. Complete the payment and return the proof-of-payment token to be used to complete the original request
```
Comment (michielbdejong) [inline letter-shop/index.md]: it doesn't just return the token, it prints complete 'go to this url to get your letter' instructions

File: letter-shop/index.md [hunk 4]
```diff
@@ -0,0 +1,634 @@
...
+
+#### Git
+
+```shell
+git clone https://github.com/interledger/tutorials
+```
+
+#### Download
+
+[ZIP Archive](https://github.com/interledger/tutorials/archive/master.zip)
```
Comment (michielbdejong) [inline letter-shop/index.md:79]: ha, silly that it hadn't occurred to me to include a clickable link for that :)

File: letter-shop/index.md [hunk 5]
```diff
@@ -0,0 +1,634 @@
...
+
+Assuming you have either cloned the repository, or unzipped your download, open a terminal window at the root of the project and change into the `letter-shop` directory.
+
+You should see the following files:
+
+```shell
+README.md
+package.json
```
Comment (michielbdejong) [inline letter-shop/index.md]: there's also a package-lock.json (which i think we can probably omit, actually)

File: letter-shop/index.md [hunk 6]
```diff
@@ -0,0 +1,634 @@
...
+
+### Interledger plugins
+
+Our shop, and many other components we'll build in this tutorial, use ledger plugins. A plugin is a piece of code that talks to a specific account on a specific ledger.
+
+Since Interledger connects potentially very different ledgers together, we need an abstraction layer that hides the specifics of the ledger, but exposes the interface needed to send and receive money over that ledger. That is what Interledger plugins are for. 
+
+Ledger plugins expose a common interface (the Ledger Plugin Interface) so that irrespective of which ledger your application is connected to, the way it sends and receives payments is identical.
+
+> **NOTE:** As we go through this tutorial we'll be using different functions of the Ledger Plugin Interface. You can find the reference documentation in [IL-RFC 0004](https://interledger.org/rfcs/0004-ledger-plugin-interface/).
```
Comment (michielbdejong) [inline letter-shop/index.md]: You removed the `/draft-8.html` from the hyperlink, was that intentional? I think it's quite important we quote specific versions of all the RFCs (then we can control how and when we update this tutorial to future RFC versions)

File: letter-shop/index.md [hunk 7]
```diff
@@ -0,0 +1,634 @@
...
+npm install
+```
+
+> If you get an error here make sure you have `node` installed.
+
+Then we try and start our shop server:
+
+```shell
+node shop.js
+```
```
Comment (michielbdejong) [inline letter-shop/index.md:113]: I love the "No account configured yet for the Shop. See 'plugins.js'." thing! Makes it really interactive.

File: letter-shop/index.md [hunk 8]
```diff
@@ -0,0 +1,634 @@
...
+  // Handle incoming transfers...
+
+})
+```
+
+Stop the server and restart it with the new code. Now it will use the ledger plugin to connect to the XRP Testnet Ledger. When it is connected (usually a few seconds) you'll see some logging in the console providing info about the ledger.
+
+We can get a lot of info from the plugin about the ledger and the account it is connected to using the `getInfo()` and `getAccount()` methods.
+
+> **NOTE:** The currency of the ledger is XRP and the currency scale is 6. That means that to send 1 XRP to this ledger we must send an Interledger payment with an amount of 1000000 (1 million Drops).
```
Comment (michielbdejong) [inline letter-shop/index.md:186]: Great addition! This is something every newcomer will ask, and I forgot to talk about it.

File: letter-shop/index.md [hunk 9]
```diff
@@ -0,0 +1,634 @@
...
+The first path handles requests to http://localhost:8000/ (i.e. no token in the URL) and defines the logic to generate a new letter, but to store this in memory until the customer pays for it.
+
+First we set the HTTP Response code to 402 (Payment Required). This is not necessary but it's a useful convention. We'll see later how this becomes more valuable.
+
+Then we generate the data required to request an Interledger payment from the customer.
+
+```js
+const secret = crypto.randomBytes(32) //Get a random 32-byte number as our secret
+const fulfillment = base64url(secret) //Encode it as a string so we can share it easily (put it in URLs etc.)
+const condition = base64url(crypto.createHash('sha256').update(secret).digest()) //Hash the secret to get the condition we'll use for the payment
```
Comment (michielbdejong) [inline letter-shop/index.md:301]: I updated the variable names in this code, since, it's now like this: https://github.com/interledger/tutorials/blob/master/shop.js#L52-L53

File: letter-shop/index.md [hunk 10]
```diff
@@ -0,0 +1,634 @@
...
+
+Now we have the plugin connected and we have the info we need about the account where we will be accepting payments. 
+
+:tada: Congratulations, our shop is performing step 1 of our 5 step plan!
+
+Our next job is to start up the web server that will host our Letter Shop service.
+
+Whenever a customer visits the website we'll try to find a proof-of-payment token in the URL. If the token is there we'll try to redeem it and give the customer their letter. If not we'll generate the letter and the proof-of-payment token and store these for later. We'll then generate another token that is derived from the proof-of-payment token and give this to the customer to attach to their payment so we can reconcile the payment and this original request.
```
Comment (michielbdejong) [inline letter-shop/index.md]: The "We'll then generate another token" part becomes a bit too much, I think. Maybe just keep it at "If not we'll generate the letter and the payment parameters, and store these for later."

File: letter-shop/index.md [hunk 11]
```diff
@@ -0,0 +1,634 @@
...
+
+```
+
+This snippet defines two code paths for incoming HTTP requests (well three but the first is just a hack to ignore favicon requests).
+
+The first path handles requests to http://localhost:8000/ (i.e. no token in the URL) and defines the logic to generate a new letter, but to store this in memory until the customer pays for it.
+
+First we set the HTTP Response code to 402 (Payment Required). This is not necessary but it's a useful convention. We'll see later how this becomes more valuable.
+
+Then we generate the data required to request an Interledger payment from the customer.
```
Comment (michielbdejong) [inline letter-shop/index.md]: (referring to my previous comment): "The payment parameters mentioned earlier consist of a fulfillment (random bytes) and a corresponding condition (its sha256 hash):"

File: letter-shop/index.md [hunk 12]
```diff
@@ -0,0 +1,634 @@
...
+
+```js
+const secret = crypto.randomBytes(32) //Get a random 32-byte number as our secret
+const fulfillment = base64url(secret) //Encode it as a string so we can share it easily (put it in URLs etc.)
+const condition = base64url(crypto.createHash('sha256').update(secret).digest()) //Hash the secret to get the condition we'll use for the payment
+```
+
```
Comment (michielbdejong) [inline letter-shop/index.md:303]: "These are the payment parameters mentioned earlier;"

File: letter-shop/index.md [hunk 13]
```diff
@@ -0,0 +1,634 @@
...
+The first piece of data is the fulfillment that we will release to the payer when the payment is delivered. The second is the condition we give to the payer to attach to the payment.
+
+The condition is simply a SHA-256 hash of the fulfillment, meaning nobody can derive the fulfillment from the condition but it's possible to quickly verify that the condition is a hash of the fulfillment.
+
+> **NOTE:** Conditions and fulfillments are a critical aspect of Interledger which we'll explain further when we attempt to make the payment. For now it's fine to simply understand that the fulfillment is a secret held by the payee and the condition is sent by the payer with their payment.
+
+In our shop we use the fulfillment as a proof-of-payment token so we store the letter we have generated for the customer using the fulfillment as the index.
+
+The rest of that branch oif the code simply returns the amount, ILP Address and condition back to the customer so they can pay for their request.
```
Comment (michielbdejong) [inline letter-shop/index.md]: typo: oif

File: letter-shop/index.md [hunk 14]
```diff
@@ -0,0 +1,634 @@
...
+
+The next part, `crypto.` indicates that we will be referring to a crypto currency's ledger. And finally, `xrp.` indicates that this ledger is the XRP testnet ledger. If you know the ledger prefix and the account, you can put them together to get the Interledger Address. 
+
+In this case, the Interledger address of our shop's account is `test.crypto.xrp.<ripple address>` and was output to the console when you started the server.
+
+> **NOTE:** Having a universal addressing scheme is critical if we want a ledger/network agnostic payment protocol. All payments need a destination, and while many traditional accounts have their own addressing scheme (IBANs, PANs, PayPal addresses), there is not a single universal scheme that can be used to address any payment.
+
+> To read more about ILP Addresses and how these can be derived for accounts on both traditional and new payment networks see [IL-RFC-15, draft 1](https://interledger.org/rfcs/0015-ilp-addresses/draft-1.html).
+
+If you run the server now you'll see that it now both step 1 and 2 of our 5 step program.
```
Comment (michielbdejong) [inline letter-shop/index.md]: "now you'll see that it now"

File: letter-shop/index.md [hunk 15]
```diff
@@ -0,0 +1,634 @@
...
+  // Listen for fulfillments...
+
+})
+```
+
+You'll see that the script connects to the ledger and returns similar information to that returned when we connected the shop to the same ledger using its own account credentials.
+
+But after that it simply hangs...
+
+The comments in the code probabaly gave it away, but we now need to make the payment. Replace the text `// Make payment...` with:
```
Comment (michielbdejong) [inline letter-shop/index.md]: probabaly

File: letter-shop/index.md [hunk 16]
```diff
@@ -0,0 +1,634 @@
...
+Now let's put ourselves in the shoes of the customer and attempt to buy our first letter!
+
+## Step 3: Paying for a letter
+
+As instructed on the console, open a browser window and go to http://localhost:8000.
+
+You should get a message along the lines of:
+
+```
+Please send an Interledger payment of 10 XRP to XXXXXXXXXXXXXXXXX using the condition YYYYYYYYYYYYYYYY
```
Comment (michielbdejong) [inline letter-shop/index.md]: 10 Drops, not 10 XRP

File: letter-shop/index.md [hunk 17]
```diff
@@ -0,0 +1,634 @@
...
+
+> Note how the amount is expressed in the scale of the ledger. We know this to be 6 therefor our payment of 0.00001 XRP is expressed as 10 Drops.
+
+> Also note how we are setting an expiry on the transfer we are preparing so that if the transfer is not fulfilled before that time it will roll-back.
+
+### Conditional Payments and Hash Timelocked Agreements
+
+At this point we should dig a little deeper into how transfers work in Interledger. They are a little different to a regular transfer on most traditional payment networks/ledgers in that they happen in two phases:
+1. The transfer is prepared pending either the fulfillment of a condition, or expiry of a timeout
+1. The transfer is either executed becuase the condition was fulfilled, or it rolls-back because it expired
```
Comment (michielbdejong) [inline letter-shop/index.md]: becuase

File: letter-shop/index.md [hunk 18]
```diff
@@ -0,0 +1,634 @@
...
+
+We call this arrangement, between the parties to the transfer, a Hash Timelock Agreement (HTLA). You may have come across HTLCs or Hash Timelock Contracts as used in many crypto-currency based system like Lightning. HTLAs are a generalization of this that don't prescribe how they are enforced (i.e. It may not be a contract written in the code of the ledger like in a crypto-currency ledger, it may be an agreement enforced by law or simply based on trust). 
+
+How two parties make transfers to one-another of a ledger will depend on the relationship between them and also the features of the underlying ledger. For our tutorial we are using a plugin specific to XRP Ledger, which uses [the on-ledger escrow](https://ripple.com/build/rippleapi/#transaction-types) features of that ledger. This means the parties are able to make transfers with no pre-existing relationship.
```
Comment (michielbdejong) [inline letter-shop/index.md]: "to one-another of a ledger"

File: letter-shop/index.md [hunk 19]
```diff
@@ -0,0 +1,634 @@
...
+
+```shell
+npm install -g jslint
+```
+
+Next use `curl` to POST an API call to the XRP Testnet server (replace the address with your sending address):
+
+```shell
+curl -X POST -d '{ "method": "account_objects", "params": [{"ledger_index":"validated", "account": "YOUR-SENDING-ADDRESS", "type": "escrow"}]}' https://client.altnet.rippletest.net:51234
+```
```
Comment (michielbdejong) [inline letter-shop/index.md]: Should use jslint here then, right?

File: letter-shop/index.md [hunk 20]
```diff
@@ -0,0 +1,634 @@
...
+
+*Installing `curl` is out of scope for this tutorial so if you don't already have it you can skip ahead or figure that bit out yourself.*
+
+You should get back a JSON object that shows all the transactions sent from your account. In among them should be the transaction you just created. 
+
+So our investigation has revealed that the transfer has been created on the ledger. We need to go back to our shop service and figure out why it wasn't received.
```
Comment (michielbdejong) [inline letter-shop/index.md:468]: Nice! I like how this shows that something really happened on the XRP testnet ledger, and also how in the tutorial you get one step further each time.

File: letter-shop/index.md [hunk 21]
```diff
@@ -0,0 +1,634 @@
...
+This is our first introduction to some of the events that can be raised by ledger plugins. There are a number of these and the one we are interested in is called `incoming_prepare`. This event is raised whenever a transfer is prepared on the ledger and your plugin's account is the intended recipient.
+
+As soon as yur plugin connects to the ledger it will start listening for incoming transfers and should raise this event every time one occurs.
+
+Replace `//Handle incoming transfers...` with the following code:
+
+```js
+  plugin.on('incoming_prepare', function (transfer) {
+
+    if (transfer.amount !== '10') {
```
Comment (michielbdejong) [inline letter-shop/index.md]: actually, it should only through InsufficientDestinationAmount if `parseInt(transfer.amount) < 10`

File: letter-shop/index.md [hunk 22]
```diff
@@ -0,0 +1,634 @@
...
+      console.log(`    - Payment complete`)
+    }
+  })
+```
+
+In this code we first check that the incoming transfer is for the correct amount, and if it's not we raise an ILP Error and call the `rejectIncomingTransfer()` function to reject the transfer. If there were other business rules we wanted to evaluate before accepting the transfer we could do these here too.
+
+If the transfer is correct, we extract the condition and use this to look up the fulfillment from our local store. If we can't find it then we aren't able to fufill the transfer so we reject it.
+
+> **NOTE:** There are other ways to map a fulfillment to an incoming transfer, including techniques where the fulfillment is derived from the transaction data. For an example of such a technique have a look at the [PSK protocol](https://interledger.org/rfcs/0016-pre-shared-key/).
```
Comment (michielbdejong) [inline letter-shop/index.md:537]: ".. the PSK protocol, which will be introduced in the [Streaming Payments](./streaming-payments) tutorial."

File: letter-shop/index.md [hunk 23]
```diff
@@ -0,0 +1,634 @@
...
+
+It's very cumbersome to copy and paste the condition from your browser to your command-line terminal each time you need to pay for something online, and then to copy and paste back the fulfillment from your terminal to your browser once you have paid. 
+
+For most paid services we'd expect the client to detect that a payment is required, make the payment, and then retry the request.
+
+To do this we need to make a few tweaks to the server and client.
+
+First, we need to make our responses machine readable so that the client can find the details of the payment required in the response. Recall how we used the 402 (Payment Required) HTTP response in the server when we responded to the initial request. Now we can see how this is useful for clients to help them recognise the response type.
+
+Another minor change that required is to put the payment details in a response header in a standard format that clients recognise.
```
Comment (michielbdejong) [inline letter-shop/index.md]: "change that required"

File: letter-shop/index.md [hunk 24]
```diff
@@ -0,0 +1,634 @@
...
+
+Modify the code in `shop.js` to look like this:
+
+```js
+      console.log(`    - Waiting for payment...`)
+
+      res.setHeader("Pay", `${cost} ${account} ${condition}`)
```
Comment (michielbdejong) [inline letter-shop/index.md:594]: In http-ilp the third string would be the PSK shared secret, not the one-off condition.

File: letter-shop/index.md [hunk 25]
```diff
@@ -0,0 +1,634 @@
...
+1. It hosts a web server at http://localhost:8001 which forwards any requests it receives to http://localhost:8000
+1. It intercepts the response and, if a payment is required, makes the payment
+1. After making a payment and getting back the fulfillment the proxy will retry the request and only then will it pass the response back to the original caller.
+
+In the terminal windo where you ran your client, start the proxy instead:
+
+```shell
+node ./proxy.js
+```
+Now, instead of visiting http://localhost:8000/, visit http://localhost:8001/. This will take a while to load because in the background the proxy is busy payng for your request.
```
Comment (michielbdejong) [inline letter-shop/index.md]: "payng"

File: letter-shop/index.md (hunk 10 above)
Comment (adrianhopebailie) [reply letter-shop/index.md]: I think that's an important aspect of the process (getting the condition). Will reword it to make it easier to digest.

**Review by michielbdejong — APPROVED** [review]
Great improvements! Some comments inline; main thing I'm thinking is obviously the 'proxy.js' thing at the end, which parses the "Please..." text as I had it, has gotten some criticism, and I don't really remember now why I wanted it there at the time. But the way it's now, it's introducing an important concept in a bonus item, and also it's using the condition instead of the psk secret, so I think that doesn't work either. How about if we just remove it, and leave it for the streaming-payments tutorial, where it can be made correct from the start?

File: letter-shop/index.md (hunk 12 above)
Comment (adrianhopebailie) [reply letter-shop/index.md:303]: I can't find where the phrase "payment parameters" is used earlier so I'm not sure what this means?

File: letter-shop/index.md (hunk 24 above)
Comment (adrianhopebailie) [reply letter-shop/index.md:594]: That's okay, we're still building to that. That's why I haven't said anything about HTTP ILP just showing what's possible

File: letter-shop/index.md (hunk 22 above)
Comment (adrianhopebailie) [reply letter-shop/index.md:537]: Will update when that is live

**Comment by adrianhopebailie** [comment]
> How about if we just remove it, and leave it for the streaming-payments tutorial, where it can be made correct from the start?

I think it's useful to have it there to demonstrate the direction we are going. If anything we should explain in the next tutorial why we don't use the condition.

**Comment by adrianhopebailie** [comment]
> When you cd into this repo on this branch and run npm install; npm test you see a lot of lint errors, mostly about spacing and indentation. Would be good to fix those, if possible.

Hmmm. Fixing will be a bit of a pain as I have to copy all the snippets again after they are fixed. Will log an issue to resolve later.
